proxy/proxy.go

```diff
@@ -287,7 +309,141 @@ func (p *Proxy) createListener(local, upstream string) (*listener.Listener, erro
 		_ = s.Disconnect(ctx)
 	}
 
-	return listener.New(logWith, sdWith, p.config.Network, local, p.config.Unlink, connectionHandler, shutdownHandler)
+	listener, err := listener.New(logWith, sdWith, p.config.Network, local, p.config.Unlink, connectionHandler, shutdownHandler)
+	if err != nil {
+		return nil, err
+	}
+	return &ListenerServerPair{
+		Listener: listener,
+		Server:   s,
+	}, nil
+}
+
+// Go through all listeners and health check their servers
+// This should ideally by its own class with a healthcheck strategy
+// Due to timeline pressures we'll just use this basic scheme
+// to avoid repeated loops of Timeout on the redis client
+func (p *Proxy) checkConnections() {
+	duration := time.Duration(p.config.ServerHealthCheckSec) * time.Second
+	p.log.Debug("Inside checkConnections", zap.String("duration", duration.String()))
+	for {
+		time.Sleep(duration)
+		p.log.Debug("Just woke up to check connections")
+		keys := p.getListenerKeys()
+		c := make(chan string)
+		for _, key := range keys {
+			go p.checkSingleConnection(key, c)
+		}
+		for range keys {
+			<-c
+		}
```

Comment:
The data being sent to the channel c isn't really used for anything? Is there a point for channel c?

channel c is created as an unbuffered channel here which means all the  checkSingleConnection() functions will not return completely until line 338 when the buffer is dumped (all contents received).  

It seems like you may not need each checkSingleConnection to write to a channel and the goroutine an just return right away when done?

Reply:
This channel is for the main routine to wait for all inidividual `checkSingleConnection`s routines to finish before starting a new loop. So we don't want any `checkSingleConnection` to finish unless the spawning routine has got a chance to OK that. 
If there's a better way to synchronize go routines please suggest since I'm not a go expert.

Reply:
Maybe what you are looking for is a [WaitGroup](https://gobyexample.com/waitgroups)?